sensor/common/compliance/nodescan_handler_test.go

```diff
@@ -51,85 +48,138 @@ func fakeNodeScanV2(nodeName string) *storage.NodeScanV2 {
+	// The intent here it to test whether the handler can shutdown with no leaks when the context is canceled despite of
+	// multiple messages waiting in the channel to be processed
 	go func() {
-		defer close(s.toCentral)
-		for {
+		defer close(nodeScans)
+		for i := 0; i < 30; i++ {
 			select {
 			case <-s.ctx.Done():
 				return
-			case msg, ok := <-c:
-				if !ok {
-					return
+			case nodeScans <- fakeNodeScanV2("Node"):
+				if i == 0 {
+					h.Stop(nil)
 				}
-				s.toCentral <- msg
 			}
 		}
 	}()
+	s.NoError(h.Start())
 }
 
-// TestStopHandler goal is to stop handler while there are still some messages to process
-// in the channel passed into NewNodeScanHandler.
-// We expect that premature stop of the handler produces no race condition or goroutine leak.
-// Exec with: go test -race -count=1 -v -run ^TestNodeScanHandler$ ./sensor/common/compliance
-func (s *NodeScanHandlerTestSuite) TestStopHandler() {
-	nodeScans := make(chan *storage.NodeScanV2)
-	h := NewNodeScanHandler(nodeScans)
-	// consume all messages send toCentral
+func (s *NodeScanHandlerTestSuite) TestHandlerRegularRoutine() {
+	h := NewNodeScanHandler(s.generateTestInput())
+	s.NoError(h.Start())
+	s.consumeToCentral(h)
+	h.Stop(nil)
+}
+
+func (s *NodeScanHandlerTestSuite) TestHandlerStoppedError() {
+	h := NewNodeScanHandler(s.generateTestInput())
+	s.NoError(h.Start())
+	s.consumeToCentral(h)
+	errTest := errors.New("example-stop-error")
+	h.Stop(errTest)
+	err := h.Stopped().Wait()
+	// if generateTestInput finishes before call to Stop(), err is errInputChanClosed
+	// otherwise it is errTest. Both are fine as a reason for stopping the handler
+	s.True(errors.Is(err, errTest) || errors.Is(err, errInputChanClosed))
+}
+
+// generateTestInput generates 10 input messages to the NodeScanHandler
+func (s *NodeScanHandlerTestSuite) generateTestInput() <-chan *storage.NodeScanV2 {
+	input := make(chan *storage.NodeScanV2)
+	// this is a producer that sends 10 nodescan messages
 	go func() {
-		for {
+		defer close(input)
+		for i := 0; i < 10; i++ {
 			select {
 			case <-s.ctx.Done():
 				return
-			case <-s.toCentral:
+			case input <- fakeNodeScanV2("Node"):
 			}
 		}
 	}()
-	s.startForwardingToCentral(h.ResponsesC())
-	// this is a producer that stops the handler after producing the first message and then sends 3 more messages
+	return input
+}
+
+// consumeToCentral starts handler and blocks until it stops
+func (s *NodeScanHandlerTestSuite) consumeToCentral(han NodeScanHandler) {
+	// simulate Central: consume all messages from h.ResponsesC()
 	go func() {
-		defer close(nodeScans)
-		for i := 0; i < 4; i++ {
+		for {
 			select {
 			case <-s.ctx.Done():
 				return
-			default:
-				nodeScans <- fakeNodeScanV2("Node")
-				if i == 0 {
-					h.Stop(nil)
+			case _, ok := <-han.ResponsesC():
+				if !ok {
+					return
 				}
 			}
 		}
 	}()
-	err := h.Start()
-	assert.NoError(s.T(), err)
 }
 
 func (s *NodeScanHandlerTestSuite) TestRestartHandler() {
```

Comment:
This and further tests should also make sure that the handler is stopped at the end.